File: sdks/node/src/files.ts
import { readFile } from 'node:fs/promises';
import { basename, extname } from 'node:path';

/**
 * Arquivo para `uploadDocument` (multipart, campo `file`). A API não aceita
 * upload por URL.
 */
export interface FileUpload {
    content: Uint8Array | ArrayBuffer | string;
    filename: string;
    contentType?: string;
}

const TYPES: Record<string, string> = {
    '.pdf': 'application/pdf',
    '.docx':
        'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    '.doc': 'application/msword',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
};

/** Lê um arquivo do disco para enviar com `uploadDocument`. */
export async function fileFromPath(
    path: string,
    contentType?: string,
): Promise<FileUpload> {
    const content = new Uint8Array(await readFile(path));
    return {
        content,
        filename: basename(path),
        contentType:
            contentType ??
            TYPES[extname(path).toLowerCase()] ??
            'application/octet-stream',
    };
}

export function toBytes(
    content: Uint8Array | ArrayBuffer | string,
): Uint8Array {
    if (typeof content === 'string') {
        return new TextEncoder().encode(content);
    }
    return content instanceof Uint8Array ? content : new Uint8Array(content);
}

/** Arquivo baixado de `downloadFile` (original, assinado ou evidências). */
export class DownloadedFile {
    constructor(
        readonly content: Uint8Array,
        readonly contentType: string | null,
        readonly filename: string | null,
    ) {}

    static fromResponse(
        headers: Record<string, string>,
        body: Uint8Array,
    ): DownloadedFile {
        const disposition = headers['content-disposition'] ?? '';
        const extended = /filename\*\s*=\s*UTF-8''([^;]+)/i.exec(disposition);
        const quoted = /filename\s*=\s*"([^"]*)"/.exec(disposition);
        const plain = /filename\s*=\s*([^;]+)/.exec(disposition);
        let filename: string | null = null;
        if (extended?.[1] !== undefined) {
            filename = decodeURIComponent(extended[1].trim());
        } else if (quoted?.[1] !== undefined) {
            filename = quoted[1].trim();
        } else if (plain?.[1] !== undefined) {
            filename = plain[1].trim();
        }
        const type = headers['content-type'];
        return new DownloadedFile(
            body,
            type === undefined ? null : (type.split(';')[0] ?? '').trim(),
            filename,
        );
    }
}

File: sdks/python/assinavelox/files.py
"""Arquivos enviados (upload multipart) e recebidos (download)."""

from __future__ import annotations

import mimetypes
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Mapping
from urllib.parse import unquote


@dataclass(frozen=True)
class FileUpload:
    """Arquivo para `upload_document` (multipart, campo `file`). A API não aceita upload por URL."""

    content: bytes
    filename: str
    content_type: str = "application/octet-stream"

    @classmethod
    def from_path(cls, path: str | Path, content_type: str | None = None) -> "FileUpload":
        file = Path(path)
        guessed = content_type or mimetypes.guess_type(file.name)[0] or "application/octet-stream"
        return cls(content=file.read_bytes(), filename=file.name, content_type=guessed)


@dataclass(frozen=True)
class DownloadedFile:
    """Arquivo baixado de `download_file` (original, assinado ou evidências)."""

    content: bytes
    content_type: str | None
    filename: str | None

    def save(self, path: str | Path) -> Path:
        target = Path(path)
        target.write_bytes(self.content)
        return target

    @classmethod
    def from_response(cls, headers: Mapping[str, str], body: bytes) -> "DownloadedFile":
        lowered = {k.lower(): v for k, v in headers.items()}
        disposition = lowered.get("content-disposition", "")
        filename = None
        extended = re.search(r"filename\*\s*=\s*UTF-8''([^;]+)", disposition, re.IGNORECASE)
        plain = re.search(r'filename\s*=\s*"([^"]*)"', disposition) or re.search(r"filename\s*=\s*([^;]+)", disposition)
        if extended:
            filename = unquote(extended.group(1).strip())
        elif plain:
            filename = plain.group(1).strip()
        content_type = lowered.get("content-type")
        return cls(content=body, content_type=content_type.split(";")[0].strip() if content_type else None, filename=filename)
